provider/bidengine/order.go

```diff
@@ -74,35 +76,75 @@ func newOrderInternal(svc *service, oid mtypes.OrderID, bid *mtypes.Bid, pricing
 	return order, nil
 }
 
-func (o *order) run() {
+var matchBidNotFound = regexp.MustCompile("^.+bid not found.+$")
+
+func (o *order) run(checkForExistingBid bool) {
 	defer o.lc.ShutdownCompleted()
+	// TODO -get context from parent
 	ctx, cancel := context.WithCancel(context.Background())
 
 	var (
 		// channels for async operations.
-		groupch   <-chan runner.Result
-		clusterch <-chan runner.Result
-		bidch     <-chan runner.Result
-		pricech   <-chan runner.Result
+		groupch       <-chan runner.Result
+		storedGroupCh <-chan runner.Result
+		clusterch     <-chan runner.Result
+		bidch         <-chan runner.Result
+		pricech       <-chan runner.Result
+		queryBidCh    <-chan runner.Result
 
 		group       *dtypes.Group
 		reservation ctypes.Reservation
 
 		won bool
+		msg *mtypes.MsgCreateBid
 	)
 
 	// Begin fetching group details immediately.
 	groupch = runner.Do(func() runner.Result {
-		res, err := o.session.Client().Query().Group(context.Background(), &dtypes.QueryGroupRequest{ID: o.orderID.GroupID()})
+		res, err := o.session.Client().Query().Group(ctx, &dtypes.QueryGroupRequest{ID: o.orderID.GroupID()})
 		return runner.NewResult(res.GetGroup(), err)
 	})
 
+	// Load existing bid if needed
+	if checkForExistingBid {
+		queryBidCh = runner.Do(func() runner.Result {
+			return runner.NewResult(o.session.Client().Query().Bid(
+				ctx,
+				&mtypes.QueryBidRequest{
+					ID: mtypes.MakeBidID(o.orderID, o.session.Provider().Address()),
+				},
+			))
+		})
+		// Hide the group details result for later
+		storedGroupCh = groupch
+		groupch = nil
+	}
 loop:
 	for {
 		select {
 		case <-o.lc.ShutdownRequest():
 			break loop
 
+		case queryBid := <-queryBidCh:
+			err := queryBid.Error()
+			bidFound := true
+			if err != nil {
+				if matchBidNotFound.MatchString(err.Error()) {
+					bidFound = false
+				} else {
+					o.session.Log().Error("could not get existing bid", "err", err, "errtype", fmt.Sprintf("%T", err))
+					break
+				}
+			}
+
+			if bidFound {
+				o.bidPlaced = true
+				_ = queryBid.Value().(*mtypes.QueryBidResponse) // TODO do I need me?
```

Comment:
no.

Reply:
Yeah I was leaning to that, I'll kill this line.